armory/datasets/preprocessing.py

```diff
@@ -57,6 +58,54 @@ def carla_over_obj_det_dev(element, modality="rgb"):
     )
 
 
+def carla_video_tracking_preprocess(element, max_frames, split):
+    x = element["video"]  # tf.Tensor [B, H, W, C]
+    y = element["bboxes"]  # tf.Tensor [B, 4]
+    y_meta = element["patch_metadata"]  # Dict
+    context = contexts[f"carla_video_tracking_{split}"]
+    # Clip
+    if max_frames:
+        max_frames = int(max_frames)
+        if max_frames <= 0:
+            raise ValueError(f"max_frames {max_frames} must be > 0")
+        x = x[:, :max_frames]
+        y = y[:, :max_frames, :]
+        y_meta = {k: v[:, :max_frames, ::] for (k, v) in y_meta.items()}
+    # Validate input
+    if x.dtype != context.input_type:
+        if x.dtype == object:
+            raise NotImplementedError(
+                "<object> dtype not yet supported for variable image processing."
+            )
+        raise ValueError(f"input dtype {x.dtype} not in ({context.input_type}, 'O')")
+    assert isinstance(
+        context.x_shape, tuple
+    ), f"target shape {context.x_shape} is not a tuple"
+    assert len(x.shape) == len(context.x_shape)
+    for a, t in zip(x.shape, context.x_shape):
+        assert (
+            t is None or a == t
+        ), f"shape {x.shape} does not match shape {context.x_shape}"
+    assert x.dtype == context.input_type
+    # Update labels
+    box_array = tf.squeeze(y, axis=0) if y.shape[0] == 1 else y
+    y = {"boxes": box_array}
+    y_meta = {
+        k: (tf.squeeze(v, axis=0) if v.shape[0] == 1 else v) for k, v in y_meta.items()
+    }
+    return x, (y, y_meta)
+
+
+@register
+def carla_video_tracking_dev(element, max_frames=None):
+    return carla_video_tracking_preprocess(element, max_frames, "dev")
+
+
+@register
+def carla_video_tracking_test(element, max_frames=None):
+    return carla_video_tracking_preprocess(element, max_frames, "test")
```

Comment:
```suggestion
def carla_video_tracking_preprocess(x, max_frames, context):
    # Clip
    if max_frames:
        max_frames = int(max_frames)
        if max_frames <= 0:
            raise ValueError(f"max_frames {max_frames} must be > 0")
        x = x[:max_frames, :]
    # Validate input
    if x.dtype != context.input_type:
        if x.dtype == object:
            raise NotImplementedError(
                "<object> dtype not yet supported for variable image processing."
            )
        raise ValueError(f"input dtype {x.dtype} not in ({context.input_type}, 'O')")
    assert isinstance(
        context.x_shape, tuple
    ), f"target shape {context.x_shape} is not a tuple"
    assert len(x.shape) == len(context.x_shape)
    for a, t in zip(x.shape, context.x_shape):
        assert (
            t is None or a == t
        ), f"shape {x.shape} does not match shape {context.x_shape}"
    assert x.dtype == context.input_type
    return x


def carla_video_tracking_preprocess_labels(y, y_patch_metadata, max_frames):
    # Clip
    if max_frames:
        max_frames = int(max_frames)
        if max_frames <= 0:
            raise ValueError(f"max_frames {max_frames} must be > 0")
        y = y[:max_frames, :]
        y_patch_metadata = {k: v[:max_frames, :] for (k, v) in y_patch_metadata.items()}
    # Update labels
    box_array = tf.squeeze(y, axis=0) if y.shape[0] == 1 else y
    y = {"boxes": box_array}
    y_patch_metadata = {
        k: (tf.squeeze(v, axis=0) if v.shape[0] == 1 else v)
        for k, v in y_patch_metadata.items()
    }
    return y, y_patch_metadata


@register
def carla_video_tracking_dev(element, max_frames=None):
    return carla_video_tracking_preprocess(
        element["video"],
        max_frames=max_frames,
        context=contexts["carla_video_tracking_dev"],
    ), carla_video_tracking_preprocess_labels(
        element["bboxes"], element["patch_metadata"], max_frames=max_frames
    )


@register
def carla_video_tracking_test(element, max_frames=None):
    return carla_video_tracking_preprocess(
        element["video"],
        max_frames=max_frames,
        context=contexts["carla_video_tracking_test"],
    ), carla_video_tracking_preprocess_labels(
        element["bboxes"], element["patch_metadata"], max_frames=max_frames
    )
```

Reply:
Addressing https://github.com/twosixlabs/armory/pull/1836/files/f07ae034e7a4415bf0bfeb3c543133bdf4e7b194#r1055888776

Reply:
See https://github.com/jprokos26/armory/commit/612fe1e7dea3880bce13e3832c592fbe5fe3df17 for a full commit.

Reply:
Accepted with https://github.com/twosixlabs/armory/pull/1836/commits/612fe1e7dea3880bce13e3832c592fbe5fe3df17.